# HW 2 Part 1: Association Rule Mining on Bike Rentals

The original dataset is available here: https://s3.amazonaws.com/baywheels-data/index.html 
We have a filtered subset of this dataset for this homework. 

The dataset is preprocessed to include all the locations from which rides on a day and hour started in 2021. If you completed HW 1 part 3, you can extract the same from your own code. The notebook below also contains code to preprocess is further for association rule mining. 

Perform the following tasks on the provided bike rental dataset. Complete all the [LP] questions to receive a "low pass" grade on the homework. The [HP] question specifications combine all parts of the homework. Please refer to Canvas for more details.

Note that if you are unable to complete any of the LP questions satisfactorily, you will receive a grade of "revision required". You can revise and resubmit your work in exchange for a token. Please review the syllabus for more information on specifications grading.

**VERY IMPORTANT**: Include **ALL** the references you used for this assignment, including names of classmates you discuss with. Failure to cite your sources counts as an act of academic dishonesty and will be taken seriously without zero tolerance. You will automatically receive a “fail” grade in the homework and further serious penalties may be imposed.

NOTE: You can look for help on the Internet but refrain from referencing too much. Please cite all your sources in your submission. 
When you submit your assignment, you automatically agree to the following statement. If you do not agree, it is your responsibility to provide the reason.

“*I affirm that I have neither given nor received unauthorized help in completing this homework. I am not aware of others receiving such help. I have cited all the sources in the solution file.*”

In [41]:
import pandas as pd
## Replace path with your file's path
path = "/content/baywheels_ride_locations_2021.csv"

sample_df = pd.read_csv(path)
sample_df



Unnamed: 0  start_date  start_hour  \
0              0  2021-01-01           0   
1              1  2021-01-01           1   
2              2  2021-01-01           2   
3              3  2021-01-01           3   
4              4  2021-01-01           4   
...          ...         ...         ...   
8743        8743  2021-12-31          19   
8744        8744  2021-12-31          20   
8745        8745  2021-12-31          21   
8746        8746  2021-12-31          22   
8747        8747  2021-12-31          23   

                                     start_station_list  
0     ['Folsom St at 3rd St', '24th St at Bartlett S...  
1     ['Howard St at 6th St', 'Hyde St at Post St', ...  
2     ['30th St at San Jose Ave', 'The Embarcadero a...  
3     ['Valencia St at 22nd St', 'Bay St at Fillmore...  
4     ['Alameda St at Henry Adams St', 'Powell St BA...  
...                                                 ...  
8743  ['Pier 1/2 at The Embarcadero', 'Pier 1/2 at T...  
8744  ['Pier 1/2 at The Embarcadero', 'El Embarcader...  
8745  ['Haight St at Lyon St', 'North Point St at Gr...  
8746  ['Scott St at Golden Gate Ave', 'Jackson St at...  
8747  ['Taylor St at 9th St', 'San Fernando St at 4t...  

[8748 rows x 4 columns]

In [42]:
stations_imported = pd.read_csv(path, index_col=0, converters={"start_station_list": lambda x: x.strip("[]").replace("'","").replace("\"","").split(", ")})
stations_imported.info()
stations_imported


<class 'pandas.core.frame.DataFrame'>
Int64Index: 8748 entries, 0 to 8747
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   start_date          8748 non-null   object
 1   start_hour          8748 non-null   int64 
 2   start_station_list  8748 non-null   object
dtypes: int64(1), object(2)
memory usage: 273.4+ KB


start_date  start_hour  \
0     2021-01-01           0   
1     2021-01-01           1   
2     2021-01-01           2   
3     2021-01-01           3   
4     2021-01-01           4   
...          ...         ...   
8743  2021-12-31          19   
8744  2021-12-31          20   
8745  2021-12-31          21   
8746  2021-12-31          22   
8747  2021-12-31          23   

                                     start_station_list  
0     [Folsom St at 3rd St, 24th St at Bartlett St, ...  
1     [Howard St at 6th St, Hyde St at Post St, Town...  
2     [30th St at San Jose Ave, The Embarcadero at P...  
3     [Valencia St at 22nd St, Bay St at Fillmore St...  
4     [Alameda St at Henry Adams St, Powell St BART ...  
...                                                 ...  
8743  [Pier 1/2 at The Embarcadero, Pier 1/2 at The ...  
8744  [Pier 1/2 at The Embarcadero, El Embarcadero a...  
8745  [Haight St at Lyon St, North Point St at Grant...  
8746  [Scott St at Golden Gate Ave, Jackson St at Po...  
8747  [Taylor St at 9th St, San Fernando St at 4th S...  

[8748 rows x 3 columns]

In [43]:
s = stations_imported['start_station_list'].explode()
s

b = stations_imported[['start_date', 'start_hour']].join(pd.crosstab(s.index, s, margins=True))
b


start_date  start_hour  10th Ave at E 15th St  10th Ave at Irving St  \
0     2021-01-01           0                      0                      0   
1     2021-01-01           1                      0                      0   
2     2021-01-01           2                      0                      0   
3     2021-01-01           3                      0                      0   
4     2021-01-01           4                      0                      0   
...          ...         ...                    ...                    ...   
8743  2021-12-31          19                      0                      0   
8744  2021-12-31          20                      0                      0   
8745  2021-12-31          21                      0                      2   
8746  2021-12-31          22                      0                      0   
8747  2021-12-31          23                      0                      0   

      10th St at Empire St  10th St at Fallon St  10th St at Mission St  \
0                        0                     0                      0   
1                        0                     0                      0   
2                        0                     0                      0   
3                        0                     0                      0   
4                        0                     0                      0   
...                    ...                   ...                    ...   
8743                     0                     0                      0   
8744                     0                     0                      0   
8745                     0                     0                      0   
8746                     0                     0                      0   
8747                     0                     0                      0   

      10th St at University Ave  11th St at Bryant St  11th St at Natoma St  \
0                             0                     0                     0   
1                             0                     0                     0   
2                             0                     0                     0   
3                             0                     0                     0   
4                             0                     0                     0   
...                         ...                   ...                   ...   
8743                          0                     0                     1   
8744                          0                     0                     0   
8745                          0                     1                     0   
8746                          0                     0                     1   
8747                          0                     0                     0   

      ...  West St at 40th St  West St at University Ave  \
0     ...                   0                          0   
1     ...                   0                          0   
2     ...                   0                          0   
3     ...                   0                          0   
4     ...                   0                          0   
...   ...                 ...                        ...   
8743  ...                   1                          0   
8744  ...                   0                          0   
8745  ...                   0                          1   
8746  ...                   0                          0   
8747  ...                   0                          0   

      William St at 10th St  Williams Ave at 3rd St  \
0                         0                       0   
1                         0                       0   
2                         0                       0   
3                         0                       0   
4                         0                       0   
...                     ...                     ...   
8743                      0                       0   
8744                      0                       0   
874

In [44]:
basket = b.drop(columns=['start_date', 'start_hour', 'All'])
basket.info()
basket


<class 'pandas.core.frame.DataFrame'>
Index: 8748 entries, 0 to 8747
Columns: 482 entries, 10th Ave at E 15th St to Yerba Buena Center for the Arts (Howard St at 3rd St)
dtypes: int64(482)
memory usage: 32.5+ MB


10th Ave at E 15th St  10th Ave at Irving St  10th St at Empire St  \
0                         0                      0                     0   
1                         0                      0                     0   
2                         0                      0                     0   
3                         0                      0                     0   
4                         0                      0                     0   
...                     ...                    ...                   ...   
8743                      0                      0                     0   
8744                      0                      0                     0   
8745                      0                      2                     0   
8746                      0                      0                     0   
8747                      0                      0                     0   

      10th St at Fallon St  10th St at Mission St  10th St at University Ave  \
0                        0                      0                          0   
1                        0                      0                          0   
2                        0                      0                          0   
3                        0                      0                          0   
4                        0                      0                          0   
...                    ...                    ...                        ...   
8743                     0                      0                          0   
8744                     0                      0                          0   
8745                     0                      0                          0   
8746                     0                      0                          0   
8747                     0                      0                          0   

      11th St at Bryant St  11th St at Natoma St  13th St at Franklin St  \
0                        0                     0                       0   
1                        0                     0                       0   
2                        0                     0                       0   
3                        0                     0                       0   
4                        0                     0                       0   
...                    ...                   ...                     ...   
8743                     0                     1                       0   
8744                     0                     0                       0   
8745                     1                     0                       0   
8746                     0                     1                       0   
8747                     0                     0                       0   

      13th St at Webster St  ...  West Oakland BART Station  \
0                         0  ...                          0   
1                         0  ...                          0   
2                         0  ...                          0   
3                         0  ...                          0   
4                         0  ...                          0   
...                     ...  ...                        ...   
8743                      0  ...                          0   
8744                      0  ...                          0   
8745                      0  ...                          0   
8746                      0  ...                          0   
8747                      0  ...                          0   

      West St at 40th St  West St at University Ave  William St at 10th St  \
0                      0                          0                      0   
1                      0                          0                      0   
2                      0                          0                      0   
3                      0                          0                      0   
4                      0                          0                     

In [45]:
for item in basket:
  basket[item] = basket[item].astype('bool')
basket
 

10th Ave at E 15th St  10th Ave at Irving St  10th St at Empire St  \
0                     False                  False                 False   
1                     False                  False                 False   
2                     False                  False                 False   
3                     False                  False                 False   
4                     False                  False                 False   
...                     ...                    ...                   ...   
8743                  False                  False                 False   
8744                  False                  False                 False   
8745                  False                   True                 False   
8746                  False                  False                 False   
8747                  False                  False                 False   

      10th St at Fallon St  10th St at Mission St  10th St at University Ave  \
0                    False                  False                      False   
1                    False                  False                      False   
2                    False                  False                      False   
3                    False                  False                      False   
4                    False                  False                      False   
...                    ...                    ...                        ...   
8743                 False                  False                      False   
8744                 False                  False                      False   
8745                 False                  False                      False   
8746                 False                  False                      False   
8747                 False                  False                      False   

      11th St at Bryant St  11th St at Natoma St  13th St at Franklin St  \
0                    False                 False                   False   
1                    False                 False                   False   
2                    False                 False                   False   
3                    False                 False                   False   
4                    False                 False                   False   
...                    ...                   ...                     ...   
8743                 False                  True                   False   
8744                 False                 False                   False   
8745                  True                 False                   False   
8746                 False                  True                   False   
8747                 False                 False                   False   

      13th St at Webster St  ...  West Oakland BART Station  \
0                     False  ...                      False   
1                     False  ...                      False   
2                     False  ...                      False   
3                     False  ...                      False   
4                     False  ...                      False   
...                     ...  ...                        ...   
8743                  False  ...                      False   
8744                  False  ...                      False   
8745                  False  ...                      False   
8746                  False  ...                      False   
8747                  False  ...                      False   

      West St at 40th St  West St at University Ave  William St at 10th St  \
0                  False                      False                  False   
1                  False                      False                  False   
2                  False                      False                  False   
3                  False                      False                  False   
4                  False                      False                  Fal

In [8]:
# Recap:
# s = stations_imported['start_station_list'].explode()
# b = stations_imported[['start_date', 'start_hour']].join(pd.crosstab(s.index, s))
# basket = b.drop(columns=['start_date', 'start_hour'])
# for col in basket:
#   basket[col] = basket[col].astype('bool')
# basket



In [46]:
## Your code goes here. Import the csv into a pandas dataframe here.
import pandas as pd
df = pd.read_csv("/content/baywheels_ride_locations_2021.csv", header=None, na_values="?")
df.shape 


(8749, 4)

[LP] Perform Association Rule Mining using the apriori algorithm on the `basket` dataframe with a support above 0.28 and confidence above 0.9.

The library you can use is mlxtend. The imports are included below. Answer the questions listed below after you perform the apriori algorithm. Answer ALL questions correctly to pass this question.

1.   Q1. The number of frequent itemsets of length 1: 
2.   Q2. The number of frequent itemsets of length 4: 
3.   Q3. The number of frequent itemsets of length 5: 
4.   Q4. How many association rules meet the confidence requirement:





In [47]:
## Your code for LP goes here. 
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

df.head(5)


0           1           2  \
0  NaN  start_date  start_hour   
1  0.0  2021-01-01          00   
2  1.0  2021-01-01          01   
3  2.0  2021-01-01          02   
4  3.0  2021-01-01          03   

                                                   3  
0                                 start_station_list  
1  ['Folsom St at 3rd St', '24th St at Bartlett S...  
2  ['Howard St at 6th St', 'Hyde St at Post St', ...  
3  ['30th St at San Jose Ave', 'The Embarcadero a...  
4  ['Valencia St at 22nd St', 'Bay St at Fillmore...

In [11]:
# This code is manually performed in upper sections which converts column data of type set including set as data and converts it to the output datafram of boolian values of each transaction occurence.

# from mlxtend.preprocessing import TransactionEncoder

# te = TransactionEncoder()
# te_ary = te.fit(basket).transform(basket)

# dff = pd.DataFrame(te_ary, columns=te.columns_)
# dff


In [48]:
ap_df = apriori(basket, min_support = 0.28, use_colnames = False)
ap_df

support                  itemsets
0     0.309442                       (1)
1     0.356996                       (6)
2     0.370713                       (7)
3     0.308871                      (13)
4     0.301440                      (16)
...        ...                       ...
3719  0.290009      (368, 449, 308, 367)
3720  0.287380      (368, 405, 310, 367)
3721  0.285094      (368, 449, 310, 367)
3722  0.285780      (368, 449, 405, 367)
3723  0.281093  (367, 368, 303, 308, 22)

[3724 rows x 2 columns]

In [49]:
ap_df = apriori(basket, min_support = 0.28, use_colnames = True)
ap_df

support                                           itemsets
0     0.309442                            (10th Ave at Irving St)
1     0.356996                             (11th St at Bryant St)
2     0.370713                             (11th St at Natoma St)
3     0.308871                            (14th St at Mission St)
4     0.301440                   (16th St Mission BART Station 2)
...        ...                                                ...
3719  0.290009  (Powell St BART Station (Market St at 4th St),...
3720  0.287380  (Market St at Steuart St, Powell St BART Stati...
3721  0.285094  (Market St at Steuart St, Powell St BART Stati...
3722  0.285780  (Powell St BART Station (Market St at 4th St),...
3723  0.281093  (Powell St BART Station (Market St at 5th St),...

[3724 rows x 2 columns]

In [50]:
freq1_count = 0; freq2_count = 0;freq3_count = 0; freq4_count = 0; freq5_count = 0
ap_df_itemsets = ap_df.itemsets
for itemset in ap_df_itemsets:
  k = len(itemset)
  if k == 1:
    freq1_count +=1
  elif k == 2:
    freq2_count +=1
  elif k == 3:
    freq3_count +=1
  elif k == 4:
    freq4_count +=1
  elif k == 5:
    freq5_count +=1
print("\n Frequency count having 1 item = {} \n Frequency count having 2 item = {} \n Frequency count having 3 item = {} \n Frequency count having 4 items = {} \n Frequency count having 5 items = {} \n".format(freq1_count,freq2_count,freq3_count,freq4_count, freq5_count))



 Frequency count having 1 item = 142 
 Frequency count having 2 item = 1393 
 Frequency count having 3 item = 1767 
 Frequency count having 4 items = 421 
 Frequency count having 5 items = 1 



Thus, 

*   A1. The number of frequent itemsets of length 1: 142
*   A2. The number of frequent itemsets of length 4: 421
*   A3. The number of frequent itemsets of length 5: 1



In [51]:
# df_1 = ap_df(len(itemsets)==1)
ap_df_itemsets

0                                 (10th Ave at Irving St)
1                                  (11th St at Bryant St)
2                                  (11th St at Natoma St)
3                                 (14th St at Mission St)
4                        (16th St Mission BART Station 2)
                              ...                        
3719    (Powell St BART Station (Market St at 4th St),...
3720    (Market St at Steuart St, Powell St BART Stati...
3721    (Market St at Steuart St, Powell St BART Stati...
3722    (Powell St BART Station (Market St at 4th St),...
3723    (Powell St BART Station (Market St at 5th St),...
Name: itemsets, Length: 3724, dtype: object

In [52]:
rules_df = association_rules(ap_df, metric= 'confidence', min_threshold=0.9)
rules_df


antecedents  \
0                            (22nd St Caltrain Station)   
1                              (7th Ave at Cabrillo St)   
2                                (8th St at Brannan St)   
3                        (Alameda St at Henry Adams St)   
4                                (Brannan St at 7th St)   
...                                                 ...   
1152  (Powell St BART Station (Market St at 4th St),...   
1153  (San Francisco Caltrain (Townsend St at 4th St...   
1154  (San Francisco Caltrain (Townsend St at 4th St...   
1155  (Powell St BART Station (Market St at 5th St),...   
1156  (Powell St BART Station (Market St at 4th St),...   

                 consequents  antecedent support  consequent support  \
0     (Market St at 10th St)            0.376315             0.66701   
1     (Market St at 10th St)            0.362940             0.66701   
2     (Market St at 10th St)            0.376772             0.66701   
3     (Market St at 10th St)            0.341907             0.66701   
4     (Market St at 10th St)            0.379858             0.66701   
...                      ...                 ...                 ...   
1152  (Market St at 10th St)            0.331619             0.66701   
1153  (Market St at 10th St)            0.312643             0.66701   
1154  (Market St at 10th St)            0.309214             0.66701   
1155  (Market St at 10th St)            0.328418             0.66701   
1156  (Market St at 10th St)            0.295725             0.66701   

       support  confidence      lift  leverage  conviction  
0     0.340764    0.905529  1.357594  0.089758    3.524773  
1     0.326703    0.900157  1.349542  0.084619    3.335156  
2     0.339849    0.902002  1.352308  0.088539    3.397945  
3     0.308413    0.902039  1.352364  0.080358    3.399230  
4     0.344193    0.906109  1.358465  0.090824    3.546561  
...        ...         ...       ...       ...         ...  
1152  0.311157    0.938297  1.406722  0.089964    5.396677  
1153  0.294010    0.940402  1.409878  0.085474    5.587293  
1154  0.291038    0.941220  1.411104  0.084790    5.665025  
1155  0.307499    0.936304  1.403733  0.088441    5.227767  
1156  0.281093    0.950522  1.425050  0.083842    6.730048  

[1157 rows x 9 columns]

In [53]:
print("The total association rules that meet the confidence requirement of 0.9 are = {} rules".format(len(rules_df)))


The total association rules that meet the confidence requirement of 0.9 are = 1157 rules


Thus, 

*   A4: The total association rules meeting the confidence requirement of 0.9 are: 1157 rules.




[HP] Explain briefly what the frequent itemsets and association rules mean. Your answer needs to be specific for this dataset. You do not need to mention the definition of these terms. Mention how these insights can be used - e.g., by the bike rental service provider or the city.

In [54]:
## Your code for HP goes here. 

!pip install apyori
from apyori import apriori
# rules = (transactions=basket, min_support = 0.5, min_confidence = 0.8)
# results = list(rules)

rules_df.antecedents.unique()
rules_df.antecedents.describe()
rules_df.consequents.unique()
rules_df.consequents.describe()

# print(type(rules_df.antecedents))

sorted_df = rules_df.sort_values('support', ascending=True)
sorted_df

# demo1 = rules_df.loc(rules_df.antecedents.isin(1,))
# demo1

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


antecedents  \
801   (Market St at Steuart St, 17th St at Valencia ...   
381           (Esprit Park, Rhode Island St at 17th St)   
1041  (Powell St BART Station (Market St at 4th St),...   
205       (Hyde St at Post St, Beale St at Harrison St)   
1049  (Jackson St at Polk St, Market St at Dolores S...   
...                                                 ...   
713   (Powell St BART Station (Market St at 4th St),...   
63    (Powell St BART Station (Market St at 5th St),...   
574   (Powell St BART Station (Market St at 4th St),...   
62    (Powell St BART Station (Market St at 4th St),...   
705   (Powell St BART Station (Market St at 4th St),...   

                 consequents  antecedent support  consequent support  \
801   (Market St at 10th St)            0.296296             0.66701   
381   (Market St at 10th St)            0.297668             0.66701   
1041  (Market St at 10th St)            0.297325             0.66701   
205   (Market St at 10th St)            0.307499             0.66701   
1049  (Market St at 10th St)            0.297211             0.66701   
...                      ...                 ...                 ...   
713   (Market St at 10th St)            0.427984             0.66701   
63    (Market St at 10th St)            0.432327             0.66701   
574   (Market St at 10th St)            0.429813             0.66701   
62    (Market St at 10th St)            0.438957             0.66701   
705   (Market St at 10th St)            0.452103             0.66701   

       support  confidence      lift  leverage  conviction  
801   0.280064    0.945216  1.417095  0.082432    6.078247  
381   0.280064    0.940860  1.410565  0.081517    5.630565  
1041  0.280064    0.941945  1.412192  0.081745    5.735815  
205   0.280178    0.911152  1.366026  0.075074    3.747884  
1049  0.280178    0.942692  1.413311  0.081936    5.810571  
...        ...         ...       ...       ...         ...  
713   0.390832    0.913194  1.369087  0.105363    3.836049  
63    0.392890    0.908778  1.362467  0.104523    3.650347  
574   0.393690    0.915957  1.373230  0.107001    3.962164  
62    0.400663    0.912760  1.368437  0.107874    3.816965  
705   0.415981    0.920101  1.379442  0.114424    4.167649  

[1157 rows x 9 columns]

In [55]:
rules_df.consequents.unique()

array([frozenset({'Market St at 10th St'}),
       frozenset({'Powell St BART Station (Market St at 4th St)'})],
      dtype=object)

In [56]:
rules_df.consequents.value_counts()

(Market St at 10th St)                            1156
(Powell St BART Station (Market St at 4th St))       1
Name: consequents, dtype: int64

In [63]:
demo2 = apriori(basket, min_support = 0.28, use_colnames = True)
demo2_rules = association_rules(ap_df, metric= 'confidence', min_threshold=0.8)
print(demo2_rules.consequents.value_counts(), demo2_rules.consequents.unique())


(Market St at 10th St)                                                  1261
(Powell St BART Station (Market St at 4th St))                           633
(Powell St BART Station (Market St at 5th St))                           519
(17th St at Valencia St)                                                 374
(Market St at Dolores St)                                                273
(Sanchez St at 15th St)                                                  225
(Market St at Steuart St)                                                183
(Powell St BART Station (Market St at 4th St), Market St at 10th St)     100
(Union Square (Powell St at Post St))                                     68
(Berry St at 4th St)                                                      60
(Powell St BART Station (Market St at 5th St), Market St at 10th St)      24
(4th St at Mission Bay Blvd S)                                            16
(Division St at Potrero Ave)                                              13

In [30]:
demo2_rules.loc[demo2_rules.consequents.isnull()]

Empty DataFrame
Columns: [antecedents, consequents, antecedent support, consequent support, support, confidence, lift, leverage, conviction]
Index: []

< Your answer for HP goes here >

=> Frequency itemsets refer to the occurence of items as per given set.

=> Associatio rules do mean that the number of frequency of occurence of item- event- is associated with the occurence of other itmes in a given itemset. We also define this by using relations with properties like support, and confidence.

=> For bike rental service provider, we can commulate undertaking points:


*   First: There are two stations having confidence >= 0.9, for minimum support count 0.28. named 'Powell St BART Station' and 'Market St at 4th St'

1.  These 2 stations should be prioritize to manage the rush of people.
2.  Bike provides shoule set the priority service to transport dropped bikes to these locations more frequently.
3.  By redusing the confidence level, we get a list of locations in decreasing frequency of usage, Using which Bike Povider may create action plan accordingly.

*   Second: Same method can be applied for definig the list of stations having confidence >= 0.8, for minimum support count 0.28. 
1.  Now this list of stations can be used to make further decesions for improving services and their management as per user usage and requirements.


*   Finally, We may conclude that more accurate action plan will be created by the rental bike providers using such data furthermore. Which may alsoi include date and time based actions too. 



# References
Include ALL your references here.

1) https://pandas.pydata.org/docs/reference/api/pandas.crosstab.html

2) https://www.youtube.com/watch?v=ViwrdwS0pAU

3) https://www.softwaretestinghelp.com/apriori-algorithm/



# What to turn in:
1. The ipynb solution file, which includes the references
2. If you use Colab or GitHub for version control, please share a link to your notebook or GitHub repository

